src/gui/PropertyModel.cpp

```diff
@@ -0,0 +1,64 @@
+#include <PropertyModel.h>
+#include <iostream>
+PropertyModel::PropertyModel(QObject* parent) : QAbstractTableModel(parent)
+{
+
+}
+
+PropertyModel::~PropertyModel()
+{
+
+}
+
+int PropertyModel::rowCount(const QModelIndex& ) const
+{
+    return properties.size();
+}
+
+int PropertyModel::columnCount(const QModelIndex& ) const
+{
+    return 2;
+}
+
+QVariant PropertyModel::data(const QModelIndex& index, int role) const
+{
+    if (role == Qt::DisplayRole)
+    {
+        if (index.column() == 0)
+            return properties[index.row()];
+        else if (index.column() == 1)
+            return propertySource->property(properties[index.row()].toStdString().c_str());
+    }
+    return QVariant();
+}
+
+Qt::ItemFlags PropertyModel::flags(const QModelIndex& index) const
+{
+    if (index.column() == 1)
+        return QAbstractTableModel::flags(index) | Qt::ItemIsEditable;
+
+    return QAbstractTableModel::flags(index);
+}
+
+void PropertyModel::setPropertySource(QObject* propertySource_)
+{
+    propertySource = propertySource_;
+
+    if (!propertySource)
+        return;
+
+    auto prop = propertySource->property("exposedProperties");
+    if (prop.isNull() || !prop.isValid())
+        return;
+
+    for (const auto& p: prop.toStringList())
```

Comment:
I wonder whether it is possible to do here:

```c++
for (auto& p : prop.toStringList()) { properties.push_back(std::move(p)); }
```
and reuse memory from of `prop.toStringList()` items.

Reply:
It might be possible, but since most Qt structures are designed with Implicit Sharing concept in mind I don't think it's necessary:
http://doc.qt.io/qt-5/implicit-sharing.html

Reply:
Interesting! So, basically, inserting an item from Qt collection into a STL collection, actually never inserts it there? I wonder now, how they solve interoperation with futures and threads.

Reply:
Well, some kind of pointer to the real data is inserted / copied. As for working with threads, it is as safe as any other object, they describe it here:
http://doc.qt.io/qt-5/threads-modules.html#threads-and-implicitly-shared-classes

Reply:
All clear, they model std shared pointers behind the scenes. While I was reading about implicit sharing I thought about string interning (i.e. a kind of a implicit sharing) that was forbidden with C++11 that ships with a memory model.